web/srv/api_handlers.go

```diff
@@ -109,3 +116,72 @@ func (h *handler) handleApiStat(w http.ResponseWriter, req *http.Request, p http
 	}
 	renderJsonPb(w, result)
 }
+
+func (h *handler) handleApiTap(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	ws, err := websocketUpgrader.Upgrade(w, req, nil)
+	if err != nil {
+		renderJsonError(w, err, http.StatusInternalServerError)
+		return
+	}
+	defer ws.Close()
+
+	var requestParams util.TapRequestParams
+	messageType, message, err := ws.ReadMessage()
+	if messageType == websocket.TextMessage {
+		err := json.Unmarshal(message, &requestParams)
+		if err != nil {
+			ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
+			return
+		}
+	} else {
+		ws.WriteMessage(websocket.CloseMessage, []byte("MessageType not supported"))
```

Comment:
I think this else clause needs a return, if the message type is not supported. And it looks like you're also not checking the error returned by the `ws.ReadMessage()` call on line 129. I'd restructure this a bit, as follows:

```
	messageType, message, err := ws.ReadMessage()
	if err != nil {
		ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
		return
	}

	if messageType != websocket.TextMessage {
		ws.WriteMessage(websocket.CloseMessage, []byte("MessageType not supported"))
		return
	}

	var requestParams util.TapRequestParams
	err = json.Unmarshal(message, &requestParams)
	if err != nil {
		ws.WriteMessage(websocket.CloseMessage, []byte(err.Error()))
		return
	}
```